[MH-12821] Crop service

isabellakutger opened the pull request.

Commit:
crop service added

Commit:
pom.xml updated

lkiesow commented:
@isabellakutger, this seems to contain part of the mattermost pull request. Can you rebase and remove this?

isabellakutger commented:
@lkiesow , sorry I didn't notice that...

lkiesow commented on docs/guides/admin/docs/modules/cropservice.md:
```diff
@@ -0,0 +1,52 @@
+# Cropping Service
+## Installation and Configuration
```
This should not be necessary anymore once this operation becomes part of the Opencast core.

lkiesow commented on docs/guides/admin/docs/modules/cropservice.md:
```diff
@@ -0,0 +1,52 @@
+`mvn clean install -Dopencast.version=4.0 -DdeployTo=/opt/opencast`
+3. Restart your Opencast server
+
+## Workflow operation
```
This should go into `docs/guides/admin/docs/workflowoperationhandlers/` and also get an entry in `docs/guides/admin/docs/workflowoperationhandlers/index.md`. Finally, the necessary entry in `mkdocs.yml` is missing.

lkiesow commented on docs/guides/admin/docs/modules/cropservice.md:
```diff
@@ -0,0 +1,52 @@
+
+  <operations>
+
+ <operation
```
Documentation for this operation should be enough.

lkiesow commented on modules/crop-api/pom.xml:
```diff
@@ -0,0 +1,53 @@
+<packaging>bundle</packaging>
+<name>crop-api</name>
+<parent>
+    <groupId>org.opencastproject</groupId>
```
Just a minor thing but can you indent using two spaces?

lkiesow commented on modules/crop-api/src/main/java/org/opencastproject/crop/api/CropException.java:
```diff
@@ -0,0 +1,56 @@
+ */
+public class CropException extends Exception {
+    /** The serial version ui */
+    private static final long serialVersionUID = 8647839276281407394L;
```
If you explicitly specify a `serialVersionUID`, it should be unique. Otherwise, just leave it to Java to generate one.

lkiesow commented on modules/crop-api/src/main/java/org/opencastproject/crop/api/package-info.java:
```diff
@@ -0,0 +1,4 @@
+/**
+ * Created by ac129583 on 27.11.17.
```
Why add this file?

lkiesow commented on modules/crop-api/pom.xml:
```diff
@@ -0,0 +1,53 @@
+    </dependency>
+    <dependency>
+        <groupId>junit</groupId>
+        <artifactId>junit</artifactId>
```
This is never used and I'm also not so sure about most of the other dependencies. Using `mvn dependency:analyze` might help

lkiesow commented on modules/crop-ffmpeg/pom.xml:
```diff
@@ -0,0 +1,129 @@
+        <groupId>org.osgi</groupId>
+        <artifactId>org.osgi.compendium</artifactId>
+    </dependency>
+    <!-- https://mvnrepository.com/artifact/org.apache.commons/commons-lang3 -->
```
?

lkiesow commented on modules/crop-ffmpeg/pom.xml:
```diff
@@ -0,0 +1,129 @@
+    <checkstyle.skip>false</checkstyle.skip>
+</properties>
+<dependencies>
+    <!-- Matterhorn dependencies -->
```
Opencast ;-P

lkiesow commented on modules/crop-ffmpeg/pom.xml:
```diff
@@ -0,0 +1,129 @@
+    </dependency>
+    <dependency>
+        <groupId>org.opencastproject</groupId>
+        <artifactId>matterhorn-workspace-api</artifactId>
```
This does not exist anymore…

lkiesow commented on modules/crop-ffmpeg/pom.xml:
```diff
@@ -0,0 +1,129 @@
+    </dependency>
+    <dependency>
+        <groupId>org.opencastproject</groupId>
+        <artifactId>matterhorn-metadata-api</artifactId>
```
Same here…

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/endpoint/CropRestEndpoint.java:
```diff
@@ -0,0 +1,145 @@
+
+  @POST
+  @Path("")
+  @Produces(MediaType.TEXT_HTML)
```
This returns XML, not HTML, doesn't it?

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/endpoint/package-info.java:
```diff
@@ -0,0 +1,4 @@
+/**
+ * Created by ac129583 on 27.11.17.
```
Same here…

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,396 @@
+
+    /** List of available operations on jobs */
+    private enum Operation {
+        Crop
```
indentation

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,396 @@
+
+    @Override
+    public void activate(ComponentContext cc) {
+        super.activate(cc);
```
4 spaces… :-)

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,396 @@
+    public void activate(ComponentContext cc) {
+        super.activate(cc);
+        final String path = cc.getBundleContext().getProperty(FFMPEG_BINARY_CONFIG);
+        this.binary = path == null ? FFMEG_BINARY_DEFAULT : path;
```
StringUtil's [defaultIfBlank](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#defaultIfBlank-T-T-) would handle a few edge cases as well and may be preferable.

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,396 @@
+    protected Track crop(Job job, Track track) throws CropException, MediaPackageException {
+
+        if (!track.hasVideo()) {
+            logger.warn("Element {} is not a track", track);
```
Unnecessary log if you throw an error anyway.

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,396 @@
+            croppedTrack =  cropFfmpeg(mediaFile, track);
+
+
+            Track cropTrack = (Track) MediaPackageElementBuilderFactory.newInstance().newElementBuilder().newElement(Track.TYPE, track.getFlavor());
```
Please limit lines to 120 characters.

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,396 @@
+            Track croppedTrack;
+
+            File mediaFile = null;
+            URL mediaUrl = null;
```
This isn't used for anything except for logging, is it? And for that, we could use `track.getURI()` directly.

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,396 @@
+
+    protected Track cropFfmpeg(File mediafile, Track track) throws IOException, CropException {
+        String outputPath = FilenameUtils.removeExtension(mediafile.getAbsolutePath()).concat("_cropdetect.mp4");
+        String[] command = new String[] { binary, "-i", mediafile.getAbsolutePath().replaceAll(" ", "\\ "), "-vf", "cropdetect=24:16:240", "-max_muxing_queue_size", "2000", "-y", outputPath};
```
Escaping is unnecessary since you pass the arguments to the process builder as array

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,396 @@
+            return track;
+        }
+        // FFMPEG-command for cropping video
+        String crop = "crop=" + "iw-" + 2 * cropValue + ":ih:" + cropValue + ":0";
```
Why not simply use the crop filter detected and printed by cropdetect?

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,396 @@
+            } catch (IOException e) {
+                throw  new CropException("Error reading the video file in the workspace", e);
+            }
+            if (track.getDuration() == null) {
```
This makes the operation fail if no inspection has been run on it in advance. Is there any reason for that?

isabellakutger replied:
no, it was just a leftover of something I tried

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,396 @@
+        logger.debug("Configuration {}: {}", FFMPEG_BINARY_CONFIG, FFMEG_BINARY_DEFAULT);
+    }
+
+    protected Track crop(Job job, Track track) throws CropException, MediaPackageException {
```
I'm wondering why you send a serialized Track to this service. From skimming over the code it seems like a simple URI would suffice?

isabellakutger replied:
I need it to create the catalog of the output file later in this method

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,396 @@
+    }
+
+    protected Track cropFfmpeg(File mediafile, Track track) throws IOException, CropException {
+        String outputPath = FilenameUtils.removeExtension(mediafile.getAbsolutePath()).concat("_cropdetect.mp4");
```
This may break if multiple operations are run simultaneously. Please generate unique file names or (even better) use ?Workspace::get(…, true)`

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,396 @@
+            throw new CropException(e);
+        } finally {
+            IoSupport.closeQuietly(outputFileStream);
+            //logger.info("Deleted local output file at {}", outputPath);
```
can be deleted

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,396 @@
+        try {
+            Job job = serviceRegistry.createJob(JOB_TYPE, Operation.Crop.toString(), Arrays.asList(MediaPackageElementParser.getAsXml(track)),
+                    cropJobLoad);
+            return job;
```
combine to one line?

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/package-info.java:
```diff
@@ -0,0 +1,4 @@
+/**
+ * Created by ac129583 on 27.11.17.
```
…

lkiesow commented on modules/crop-workflowoperation/pom.xml:
```diff
@@ -0,0 +1,59 @@
+<dependencies>
+    <dependency>
+        <groupId>org.opencastproject</groupId>
+        <artifactId>matterhorn-common</artifactId>
```
`matterhorn-…`

lkiesow commented on modules/crop-workflowoperation/src/main/java/org/opencastproject/workflow/handler/crop/CropWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,184 @@
+     * @see org.opencastproject.workflow.api.WorkflowOperationHandler#start(org.opencastproject.workflow.api.WorkflowInstance, JobContext)
+     */
+    @Override
+    public WorkflowOperationResult start(WorkflowInstance workflowInstance, JobContext jobContext) throws WorkflowOperationException {
```
120 cols

lkiesow commented on modules/crop-workflowoperation/src/main/java/org/opencastproject/workflow/handler/crop/CropWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,184 @@
+            return createResult(WorkflowOperationResult.Action.CONTINUE);
+        }
+
+        if (candidates.size() > 1) {
```
Wouldn't it make sense to crop all instead of just picking a random one

lkiesow reviewed (changes requested):
I took a quick first glance at this and the whole thing looks good in general. I've commented on a lot of minor stuff.

One thing that's definitely missing though is the build integration: The entries in the main `pom.xml` and the `feature.xml`.

Commit:
canges according review

lkiesow commented:
@isabellakutger, please let me know if you need any assistance or tips to get this working or if you want me to do another round of reviews. Note that currently both the documentation builds and the builds of Opencast are failing. For more details, take a look at the Travis build log.

Commit:
test errors fixed

Commit:
new try to fix test errors

Commit:
a new try to fix test error

isabellakutger commented:
@lkiesow I fixed the occurring errors and changed the code as you requested in your review. Would you take a look at these changes, please?

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,424 @@
+    String outputPath = FilenameUtils.removeExtension(mediafile.getAbsolutePath()).concat(RandomStringUtils
+            .randomAlphanumeric(8) + ".mp4");
+    String[] command = new String[] { binary, "-i", mediafile.getAbsolutePath(), "-vf",
+            "cropdetect=24:16:240", "-max_muxing_queue_size", "2000", "-y", outputPath };
```
Instead of encoding a new file, you could just decode and filter it. That should be much faster:

    ffmpeg -nostats -i in.mp4 -filter:v cropdetect=24:16:240 -f null -

It would also be nice to be able to expose the cropdetect configuration though a configuration file or even use an encoding profile for this.

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,424 @@
+    } finally {
+      IoSupport.closeQuietly(process);
+      IoSupport.closeQuietly(errStream);
+      if (exitCode != 0) {
```
This part if unnecessary if you do not generate any output files.

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,424 @@
+    String croppedOutputPath = FilenameUtils.removeExtension(mediafile.getAbsolutePath()).concat(RandomStringUtils
+            .randomAlphanumeric(8) + ".mp4");
+    String[] cropCommand = new String[] { binary, "-i", mediafile.getAbsolutePath(), "-vf", crop,
+            "-max_muxing_queue_size", "2000", "-y", croppedOutputPath };
```
It would probably make sense to be able to configure the encoder options here. Possibly even using an encoding profile.

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,424 @@
+    logger.info("running {}", cropCommandline);
+
+    ProcessBuilder pbuilder1 = new ProcessBuilder(cropCommand);
+    Process process1 = pbuilder1.start();
```
Why not just re-use `process` and `pbuilder?

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,424 @@
+    ProcessBuilder pbuilder1 = new ProcessBuilder(cropCommand);
+    Process process1 = pbuilder1.start();
+    BufferedReader reader = new BufferedReader(new InputStreamReader(process1.getErrorStream()));
+    try {
```
Why read the whole input if you throw it away anyway?

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,424 @@
+    } catch (IllegalArgumentException e) {
+      throw new CropException(e);
+    } finally {
+      IoSupport.closeQuietly(outputFileStream);
```
Using a try-with-resource statement make the whole thing easier.

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,424 @@
+    logger.debug("Configuring the cropper");
+
+    cropJobLoad = LoadUtil
+            .getConfiguredLoadValue(dictionary, CROP_JOB_LOAD_KEY, DEFAULT_CROP_JOB_LOAD, serviceRegistry);
```
Can you add the configuration file for this?

lkiesow commented on modules/crop-workflowoperation/src/main/java/org/opencastproject/workflow/handler/crop/CropWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,189 @@
+      candidates.addAll(Arrays.asList(mediaPackage.getTracks(MediaPackageElements.PRESENTATION_SOURCE)));
+    }
+
+    Iterator<Track> ti = candidates.iterator();
```
Using `…stream().filter(…)…` is probably easier.

isabellakutger replied:
I was getting a ArrayIndexOutOfBoundsException when building the module

lkiesow replied:
Didn't read your code correctly, sorry. For using `filter()`, you would have to copy the list. For removing some elements, this should work and is much simpler:

```Java
candidates.removeIf(t -> !t.hasVideo());
```

lkiesow commented on modules/crop-workflowoperation/src/main/java/org/opencastproject/workflow/handler/crop/CropWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,189 @@
+
+    // all canidates are cropped in the following loop, but I'm not sure if I'm doing this in the right way
+    Job job = null;
+    for (int i = 0; i < candidates.size(); i++) {
```
`for (Track candidate: candidates)` is simpler

lkiesow commented on modules/crop-workflowoperation/src/main/java/org/opencastproject/workflow/handler/crop/CropWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,189 @@
+      Track croppedTrack = null;
+
+      try {
+        job = cropService.crop(track);
```
Don't wait for each job but start all jobs at once and then wait for all to be finished. We have a distributed cluster to work on after all.

lkiesow commented on modules/crop-workflowoperation/src/main/java/org/opencastproject/workflow/handler/crop/CropWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,189 @@
+    // all canidates are cropped in the following loop, but I'm not sure if I'm doing this in the right way
+    Job job = null;
+    for (int i = 0; i < candidates.size(); i++) {
+      Track track = candidates.get(0);
```
You want `candidates.get(i)` but using a fereach makes this unnecessary as well.

lkiesow commented on modules/crop-workflowoperation/src/main/java/org/opencastproject/workflow/handler/crop/CropWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,189 @@
+        croppedTrack.setURI(workspace
+                .moveTo(croppedTrack.getURI(), mediaPackage.getIdentifier().toString(), croppedTrack.getIdentifier(),
+                        croppedTrack.getIdentifier() + "corpped"));
+        croppedTrack.setMimeType(track.getMimeType());
```
The output mime type could be completely different from the input mime type. If you want it to be set, you can use inspection on the output track.

lkiesow reviewed (changes requested):
Did another round of reviews and found a few problems. Additionally  to these, I noticed that:

- A remote implementation is missing, making it necessary to run this on the admin node. Due to it doing video encoding, on larger clusters, this cloud hence easily kill the admin node. Hence, please add a remote module.
- It would be nice to get some unit tests for this code
- Can you add the modules to the `feature.xml` to automatically be included in the distributions?
- Can you indent the whole XML using two spaces? …just to be consistent.

Commit:
changes according review

lkiesow commented:
Note that the test failure in `silencedetection-impl` was a bug unrelated to this pull request. It has already been fixed in `develop`.

Commit:
line length in cropvideo.md changed

Commit:
trailing spaces in corpvideo-woh.mdremoved

Commit:
Test for crop-service-ffmpeg added

Commit:
Fix error in costum.properties

isabellakutger commented:
@lkiesow could you do another review, please?

lkiesow commented on assemblies/karaf-features/src/main/feature/feature.xml:
```diff
@@ -376,6 +379,9 @@
     <bundle start-level="82">mvn:org.opencastproject/opencast-workflow-service-impl/${project.version}</bundle>
     <bundle start-level="82">mvn:org.opencastproject/opencast-workflow-workflowoperation/${project.version}</bundle>
+    <bundle start-level="82">mvn:org.opencastproject/crop-api/${project.version}</bundle>
+    <bundle start-level="82">mvn:org.opencastproject/crop-ffmpeg/${project.version}</bundle>
```
This would put the admin node under heavy load since it would need to run encoding jobs. To prevent that, Opencast's encoding services usually come with a remote implementation [like this](https://github.com/opencast/opencast/blob/7c84e955364bac7f715cce1feac5ba35faf50aa1/modules/animate-remote/src/main/java/org/opencastproject/animate/remote/AnimateServiceRemoteImpl.java) which will simply accept the job and ship it to a worker node. Could you add such a remote module?

lkiesow commented on assemblies/karaf-features/src/main/feature/feature.xml:
```diff
@@ -500,6 +506,9 @@
     <bundle start-level="82">mvn:org.opencastproject/opencast-workflow-service-impl/${project.version}</bundle>
     <bundle start-level="82">mvn:org.opencastproject/opencast-workflow-workflowoperation/${project.version}</bundle>
+    <bundle start-level="82">mvn:org.opencastproject/crop-api/${project.version}</bundle>
+    <bundle start-level="82">mvn:org.opencastproject/crop-ffmpeg/${project.version}</bundle>
```
As non-worker, this should get the remote implementation as well.

lkiesow commented on assemblies/karaf-features/src/main/feature/feature.xml:
```diff
@@ -556,6 +565,10 @@
     <bundle start-level="82">mvn:org.opencastproject/opencast-waveform-ffmpeg/${project.version}</bundle>
     <bundle start-level="82">mvn:org.opencastproject/opencast-workflow-service-remote/${project.version}</bundle>
+    <!-- Fixes MH-11263 - add LTI dependency to the kernel bundle for shared opencast.httpcontext -->
+    <bundle start-level="82">mvn:org.opencastproject/opencast-lti/${project.version}</bundle>
```
I think the lit module got added by accident?

lkiesow commented on assemblies/karaf-features/src/main/feature/feature.xml:
```diff
@@ -556,6 +565,10 @@
+    <!-- Fixes MH-11263 - add LTI dependency to the kernel bundle for shared opencast.httpcontext -->
+    <bundle start-level="82">mvn:org.opencastproject/opencast-lti/${project.version}</bundle>
+    <bundle start-level="82">mvn:org.opencastproject/crop-api/${project.version}</bundle>
+    <bundle start-level="82">mvn:org.opencastproject/crop-ffmpeg/${project.version}</bundle>
```
The ingest node probably does not need the service at all, does it? The node should only be meant to accept ingests from capture agents.

lkiesow commented on assemblies/karaf-features/src/main/feature/feature.xml.orig:
```diff
@@ -0,0 +1,753 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This file was probably added by accident.

lkiesow commented on docs/guides/admin/docs/workflowoperationhandlers/cropvideo-woh.md:
```diff
@@ -0,0 +1,22 @@
+The plugin provides the workflow operation `crop-video`. This workflow operation excutes ffmpeg command "cropdetect".
+`cropdetect` checks for black bars on the sides of the track of the workflow instance. If `corptetect` is successfully,
+then ffmpeg command `crop` is executed. `crop` removes these black bars.
+
```
Can you add a description of the supported parameters like the [other operations](https://github.com/opencast/opencast/blob/7c84e955364bac7f715cce1feac5ba35faf50aa1/docs/guides/admin/docs/workflowoperationhandlers/compose-woh.md) have?

lkiesow commented on modules/crop-workflowoperation/src/main/java/org/opencastproject/workflow/handler/crop/CropWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,189 @@
+  private static final String PROP_TARGET_TAGS = "target-tags";
+
+  /** The configuration options for this handler */
+  private static final SortedMap<String, String> CONFIG_OPTIONS;
```
I think this, the `CONFIG_OPTIONS` definition below and the method `getConfigurationOptions` aren't actually needed for WOHs anymore and you could just omit them.

lkiesow commented on modules/crop-workflowoperation/src/main/java/org/opencastproject/workflow/handler/crop/CropWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,189 @@
+      logger.info("Found more than one track to crop");
+    }
+
+    // all canidates are cropped in the following loop, but I'm not sure if I'm doing this in the right way
```
Let's do it in the right way and get rid of the comment ;-D
The most important thing here is that as a first step we want to start all of the jobs at once and only afterward wait for them to finish. That way, several workers can do the cropping at once if that is necessary.

My suggested code update would look like this. Note that I haven't tested this!!

```Java
diff --git a/modules/crop-workflowoperation/src/main/java/org/opencastproject/workflow/handler/crop/CropWorkflowOperationHandler.java b/modules/crop-workflowoperation/src/main/java/org/opencastproject/workflow/handler/crop/CropWorkflowOperationHandler.java
index 765abc38d..62068d2db 100644
--- a/modules/crop-workflowoperation/src/main/java/org/opencastproject/workflow/handler/crop/CropWorkflowOperationHandler.java
+++ b/modules/crop-workflowoperation/src/main/java/org/opencastproject/workflow/handler/crop/CropWorkflowOperationHandler.java
@@ -20,6 +20,7 @@
  */
 package org.opencastproject.workflow.handler.crop;
 
+import org.opencastproject.crop.api.CropException;
 import org.opencastproject.crop.api.CropService;
 import org.opencastproject.job.api.Job;
 import org.opencastproject.job.api.JobContext;
@@ -27,9 +28,11 @@ import org.opencastproject.mediapackage.MediaPackage;
 import org.opencastproject.mediapackage.MediaPackageElementFlavor;
 import org.opencastproject.mediapackage.MediaPackageElementParser;
 import org.opencastproject.mediapackage.MediaPackageElements;
+import org.opencastproject.mediapackage.MediaPackageException;
 import org.opencastproject.mediapackage.Track;
 import org.opencastproject.mediapackage.identifier.IdBuilder;
 import org.opencastproject.mediapackage.identifier.IdBuilderFactory;
+import org.opencastproject.util.NotFoundException;
 import org.opencastproject.workflow.api.AbstractWorkflowOperationHandler;
 import org.opencastproject.workflow.api.WorkflowInstance;
 import org.opencastproject.workflow.api.WorkflowOperationException;
@@ -42,6 +45,8 @@ import org.apache.commons.lang3.StringUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.io.File;
+import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Iterator;
@@ -102,7 +107,6 @@ public class CropWorkflowOperationHandler extends AbstractWorkflowOperationHandl
           throws WorkflowOperationException {
     logger.debug("Running cropping on workflow {}", workflowInstance.getId());
 
-    final String targetTrackId = idBuilder.createNew().toString();
     WorkflowOperationInstance operation = workflowInstance.getCurrentOperation();
     MediaPackage mediaPackage = workflowInstance.getMediaPackage();
 
@@ -132,34 +136,53 @@ public class CropWorkflowOperationHandler extends AbstractWorkflowOperationHandl
       logger.info("Found more than one track to crop");
     }
 
-    // all canidates are cropped in the following loop, but I'm not sure if I'm doing this in the right way
-    long totalTimeInQueue = 0;
+    // start cropping all candidates in parallel
+    List<Job> jobs = new ArrayList<>();
     for (Track candidate: candidates) {
-      Job job;
-      Track croppedTrack = null;
-
       try {
-        job = cropService.crop(candidate);
-        if (!waitForStatus(job).isSuccess()) {
-          throw new WorkflowOperationException("Video cropping of " + candidate + " failed");
-        }
+        jobs.add(cropService.crop(candidate));
+      } catch (MediaPackageException | CropException e) {
+        throw new WorkflowOperationException("Failed starting crop job", e);
+      }
+    }
+
+    // wait for all crop jobs to be finished
+    if (!waitForStatus(jobs.toArray(new Job[0])).isSuccess()) {
+      throw new WorkflowOperationException("Crop operation failed");
+    }
+
+    long totalTimeInQueue = 0;
 
+    // add new tracks to media package
+    for (Job job: jobs) {
+      // deserialize track
+      Track croppedTrack;
+      try {
         croppedTrack = (Track) MediaPackageElementParser.getFromXml(job.getPayload());
-        mediaPackage.add(croppedTrack);
-        croppedTrack.setIdentifier(targetTrackId);
-        croppedTrack.setURI(workspace
-                .moveTo(croppedTrack.getURI(), mediaPackage.getIdentifier().toString(), croppedTrack.getIdentifier(),
-                        croppedTrack.getIdentifier() + "corpped"));
-
-        // Add target tags
-        for (String tag : targetTags) {
-          croppedTrack.addTag(tag);
-        }
-      } catch (Exception e) {
-        throw new WorkflowOperationException(e);
+      } catch (MediaPackageException e) {
+        throw new WorkflowOperationException(String.format("Crop service yielded invalid track: %s", job.getPayload()));
+      }
+
+      // update identifier
+      croppedTrack.setIdentifier(idBuilder.createNew().toString());
+
+      // move into space for media package in ws/wfr
+      try {
+        String filename = "cropped_" + new File(croppedTrack.getURI()).getName();
+        croppedTrack.setURI(workspace.moveTo(croppedTrack.getURI(), mediaPackage.getIdentifier().toString(),
+                croppedTrack.getIdentifier(), filename));
+      } catch (NotFoundException | IOException e) {
+        throw new WorkflowOperationException(String.format("Could not move %s to media package %s",
+                croppedTrack.getURI(), mediaPackage.getIdentifier()));
       }
-      long timeInQueue = job.getQueueTime() == null ? 0 : job.getQueueTime();
-      totalTimeInQueue += timeInQueue;
+
+      // Add target tags
+      targetTags.forEach(croppedTrack::addTag);
+
+      // add new track to mediapackage
+      mediaPackage.add(croppedTrack);
+
+      totalTimeInQueue += job.getQueueTime() == null ? 0 : job.getQueueTime();
     }
     logger.info("Video cropping completed");
     return createResult(mediaPackage, WorkflowOperationResult.Action.CONTINUE, totalTimeInQueue);
```

lkiesow replied:
I put that patch up at https://data.lkiesow.de/opencast/CropWorkflowOperationHandler.patch

isabellakutger replied:
Thanks:)

lkiesow commented on modules/crop-workflowoperation/pom.xml:
```diff
@@ -0,0 +1,60 @@
+    <opencast.basedir>${project.basedir}/../..</opencast.basedir>
+    <checkstyle.skip>false</checkstyle.skip>
+  </properties>
+  <dependencies>
```
`org.apache.commons:commons-lang3` and `org.slf4j:slf4j-api` seem to be used but not declared as dependencies.

lkiesow commented on docs/guides/admin/docs/releasenotes.md:
```diff
@@ -11,7 +11,7 @@ Configuration changes
 
 - The tracking options defaults have changed to be more aware of the European Union's General Data Protection
-  Regulation. Note that you can [still use the old settings if you want to](configuration/logging.and.privacy.md).
+  Regulation. Note that you can [still use the old settings if you want to](configuration/user-statistics.and.privacy.md).
```
While this is a valid fix and I don't really mind merging this as part of this pull request, it's generally preferable to create a new pull request for unrelated changes. As long as it's a simple documentation fix, you don't even need a Jira ticket. That way things like this get reviewed and fixed faster and the reviewer for the main pull request does not get confused by any unrelated changes.

lkiesow commented on modules/crop-workflowoperation/pom.xml:
```diff
@@ -0,0 +1,60 @@
+            <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+            <Build-Number>${buildNumber}</Build-Number>
+            <Export-Package>
+              org.opencastproject.workflow.handler.crop;version=${project.version
```
The missing closing bracket makes Opencast fail to start up at all:

```
2018-05-30 22:40:00,336 | ERROR | (BootFeaturesInstaller:126) - Error installing boot features
org.osgi.framework.BundleException: Could not create bundle object.
        at org.apache.felix.framework.Felix.installBundle(Felix.java:3079)[org.apache.felix.framework-5.6.4.jar:]
        ...
        at java.lang.Thread.run(Thread.java:748)[:1.8.0_171]
Caused by: java.lang.IllegalArgumentException: invalid version "project.version": non-numeric "project"
```

lkiesow commented on modules/crop-ffmpeg/pom.xml:
```diff
@@ -0,0 +1,125 @@
+        <configuration>
+          <instructions>
+            <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+            <Build-Number>${buildNumber}</Build-Number>
```
Right now, Opencast will fail to start with an OSGI wiring error if your patch is added:

```
2018-05-30 22:33:54,278 | ERROR | (BootFeaturesInstaller:126) - Error installing boot features
org.osgi.service.resolver.ResolutionException: Unable to resolve root: missing requirement [root] osgi.identity; osgi.identity=opencast-allinone; type=karaf.feature; version="[0,0.0.0]"; filter:="(&(osgi.identity=opencast-allinone)(type=karaf.feature)(version>=0.0.0)(version<=0.0.0))" [caused by: Unable to resolve opencast-allinone/0.0.0: missing requirement [opencast-allinone/0.0.0] osgi.identity; osgi.identity=crop-ffmpeg; type=osgi.bundle; version="[6.0.0.SNAPSHOT,6.0.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve crop-ffmpeg/6.0.0.SNAPSHOT: missing requirement [crop-ffmpeg/6.0.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=javax.ws.rs)(version>=1.1.0)(!(version>=2.0.0)))"]]
        at org.apache.felix.resolver.ResolutionError.toException(ResolutionError.java:42)[6:org.apache.karaf.features.core:4.0.10]
        at org.apache.felix.resolver.ResolverImpl.doResolve(ResolverImpl.java:391)[6:org.apache.karaf.features.core:4.0.10]
```

Adding this will fix the issue:

```diff
diff --git a/modules/crop-ffmpeg/pom.xml b/modules/crop-ffmpeg/pom.xml
index b8bd64138..8a45664b6 100644
--- a/modules/crop-ffmpeg/pom.xml
+++ b/modules/crop-ffmpeg/pom.xml
@@ -112,2 +112,7 @@
             <Build-Number>${buildNumber}</Build-Number>
+            <Import-Package>
+              javax.ws.rs;version=2.0.1,
+              javax.ws.rs.core;version=2.0.1,
+              *
+            </Import-Package>
             <Export-Package>
```

lkiesow commented on etc/custom.properties:
```diff
@@ -306,6 +306,10 @@ org.opencastproject.dataloader.csv=${karaf.etc}/dataimport
+######### CROP SERVICE CONFIGURATION #########
+
+# The job load for crop service
+#org.opencastproject.job.load.cropping=1.0
```
Does this work? Using `ManagedService::update`, you would usually add a configuration file for this service like 
[etc/org.opencastproject.animate.impl.AnimateServiceImpl.cfg](https://github.com/opencast/opencast/blob/7c84e955364bac7f715cce1feac5ba35faf50aa1/etc/org.opencastproject.animate.impl.AnimateServiceImpl.cfg)

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,403 @@
+    super.activate(cc);
+    final String path = cc.getBundleContext().getProperty(FFMPEG_BINARY_CONFIG);
+    this.binary = StringUtils.defaultIfBlank(path, FFMEG_BINARY_DEFAULT);
+    logger.debug("Configuration {}: {}", FFMPEG_BINARY_CONFIG, FFMEG_BINARY_DEFAULT);
```
The second parameter should probably be `binary`, should it not?

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,403 @@
+      try {
+        uri = croppedTrack.getURI();
+      } catch (Exception e) {
+        throw new CropException("Unable to put the mpeg7 catalog into the workspace", e);
```
No one tries to put anything into the workspace in this block.

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,403 @@
+
+      return cropTrack;
+
+    } catch (Exception e) {
```
I think the only thing throwing an exception in here is 

      croppedTrack = cropFfmpeg(mediaFile, track);

You could just put a `try {} catch (IOException)` around that.

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,403 @@
+
+  protected Track cropFfmpeg(File mediafile, Track track) throws IOException, CropException {
+    String[] command = new String[] { binary, "-i", mediafile.getAbsolutePath(), "-vf", "cropdetect=24:16:240",
+            "-max_muxing_queue_size", "2000", "-f", "null", "-"};
```
Still the same comment from before: This should preferably be configurable.

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,403 @@
+    pbuilder.redirectErrorStream(true);
+    int exitCode = 1;
+    Process process = pbuilder.start();
+    BufferedReader errStream = new BufferedReader(new InputStreamReader(process.getInputStream()));
```
```diff
     Process process = pbuilder.start();
-    BufferedReader errStream = new BufferedReader(new InputStreamReader(process.getInputStream()));
-    try {
+    try (BufferedReader errStream = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
       String line = errStream.readLine();
```

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,403 @@
+    } catch (IOException e) {
+      logger.error("Error executing ffmeg: {}", e.getMessage());
+    } catch (InterruptedException e) {
+      logger.error("Waiting for encoder process exited was interrupted unexpected: {}", e.getMessage());
```
`logger.error("Encoder process was unexpectedly interrupted", e);`￼

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,403 @@
+    if (exitCode != 0) {
+      throw new CropException("The encoder process exited abnormally with exit code " + exitCode);
+    }
+    if (cropValue > widthVideo / 3) {
```
magic? ;-P

lkiesow commented on modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java:
```diff
@@ -0,0 +1,403 @@
+      }
+
+    } catch (Exception e) {
+      logger.error("Error executing ffmeg: {}", e.getMessage());
```
´logger.error("Error executing ffmeg", e);`
Don't hide stack traces

lkiesow commented on modules/crop-ffmpeg/src/test/java/org/opencastproject/crop/CropServiceTest.java:
```diff
@@ -0,0 +1,209 @@
+    Workspace workspace = EasyMock.createNiceMock(Workspace.class);
+    EasyMock.expect(workspace.get((URI) EasyMock.anyObject())).andReturn(new File(track.getURI()));
+    tempFile = testFolder.newFile(getClass().getName() + ".xml");
+    EasyMock.expect(workspace.putInCollection((String) EasyMock.anyObject(), (String) EasyMock.anyObject(),
```
You can make this simpler by using lambdas and the more specific easymock methods:

```diff
diff --git a/modules/crop-ffmpeg/src/test/java/org/opencastproject/crop/CropServiceTest.java b/modules/crop-ffmpeg/src/test/java/org/opencastproject/crop/CropServiceTest.java
index 6e78e464b..02d52f0e1 100644
--- a/modules/crop-ffmpeg/src/test/java/org/opencastproject/crop/CropServiceTest.java
+++ b/modules/crop-ffmpeg/src/test/java/org/opencastproject/crop/CropServiceTest.java
@@ -120,11 +120,8 @@ public class CropServiceTest {
     tempFile = testFolder.newFile(getClass().getName() + ".xml");
-    EasyMock.expect(workspace.putInCollection((String) EasyMock.anyObject(), (String) EasyMock.anyObject(),
-            (InputStream) EasyMock.anyObject())).andAnswer(new IAnswer<URI>() {
-      @Override
-      public URI answer() throws Throwable {
-        InputStream in = (InputStream) EasyMock.getCurrentArguments()[2];
-        IOUtils.copy(in, new FileOutputStream(tempFile));
-        return tempFile.toURI();
-      }
-    });
+    EasyMock.expect(workspace.putInCollection(EasyMock.anyString(), EasyMock.anyString(),
+            EasyMock.anyObject(InputStream.class))).andAnswer(() -> {
+              InputStream in = (InputStream) EasyMock.getCurrentArguments()[2];
+              IOUtils.copy(in, new FileOutputStream(tempFile));
+              return tempFile.toURI();
+            });
     EasyMock.replay(workspace);
```

lkiesow reviewed (dismissed):
I did another review. The most important things are the two changes to the pom.xml files I posted since Opencast will not start up at all without these changes.

For binary test files you could also use this which should work as well but is ~5k instead of 600k:
https://data.lkiesow.de/opencast/black.mp4

Commit:
Changes according review

Commit:
small change in CropOperationHandler.java

Commit:
unused import removed

Commit:
trailing space removed

Commit:
crop command configurable

Commit:
trailing space removed

lkiesow committed:
Fix issues with crop service integration

Main changes:

- Use Opencast prefix for bundles
- Ensure crop service is on worker (3 server set-up)
- Actually build crop-remote (fails right now)
- Fix several dependency versions

lkiesow committed:
Simplify Crop Service

- Remove unnecessary variables
- Refrain from packaing and unpacking files in tracks

lkiesow commented:
Hi Isabella,
I did take another look at this and still found a few issues. Most notably, the remote implementation was not added to the main `pom.xml` and is hence never built. That causes the distribution steps (where it is added) to fail. If a `mvn clean install` does not fail for you, try removing your local maven cache at `~/.m2` before you start maven.

After you add the remote, you will notice that that module does not build either. I did not look into it in more detail right now.

I have two patches fixing a few of these issues. Let me know if you need any explanation to any of the changes:
- [Fixing module integration and build stuff](https://github.com/lkiesow/opencast/commit/7e60ed9c3a721fb9a2dc9746521b309528d8e7e0). These are more or less requirements. Though there are often slightly other solutions if you prefer them.
- I've also played a bit with [simplifying the crop service](https://github.com/lkiesow/opencast/commit/676fbc25588c9c609fa6cf267a02250bec3c9c9b). So far I did not test those changes due to the build error and most of them are not hard requirements. Hence, feel free to completely discard that patch if you don't like it.

If you have any questions or need help, please don't hesitate to let me know.

Commit:
build error resolved

Commit:
Fix package name

lkiesow commented:
Tried this again. It now builds and starts up just fine. Running it, I get an error which the following patch fixes:

```diff
diff --git a/modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java b/modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java
index 6e0820581..d8bebe63b 100644
--- a/modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java
+++ b/modules/crop-ffmpeg/src/main/java/org/opencastproject/crop/impl/CropServiceImpl.java
@@ -50,11 +50,14 @@ import org.slf4j.LoggerFactory;
 
 import java.io.BufferedReader;
 import java.io.File;
+import java.io.FileInputStream;
 import java.io.IOException;
 import java.io.InputStreamReader;
+import java.net.URI;
 import java.util.Arrays;
 import java.util.Dictionary;
 import java.util.List;
+import java.util.UUID;
 
 /**
  * Media analysis plugin that takes a video stream and removes black bars on each side
@@ -184,10 +187,17 @@ public class CropServiceImpl extends AbstractJobProducer implements CropService,
     logger.info("Starting cropping of {}", track);
 
     File croppedMedia = cropFFmpeg(mediaFile);
+    String fileName = UUID.randomUUID() + "-" + croppedMedia.getName();
+    URI croppedMediaUri;
+    try (FileInputStream fileStream = new FileInputStream(croppedMedia)) {
+      croppedMediaUri = workspace.putInCollection(COLLECTION_ID, fileName, fileStream);
+    } catch (IOException e) {
+      throw new CropException("Error putting output file in workspace collection", e);
+    }
 
     Track cropTrack = (Track) MediaPackageElementBuilderFactory.newInstance().newElementBuilder()
             .newElement(Track.TYPE, track.getFlavor());
-    cropTrack.setURI(croppedMedia.toURI());
+    cropTrack.setURI(croppedMediaUri);
 
     logger.info("Finished video cropping of {}", track.getURI());
     return cropTrack;
```

After that, I tried adding this to a workflow, intuitively using it like this:

```xml
    <operation
      id="crop-video"
      fail-on-error="true"
      description="Detecting black bars in presentation track">
      <configurations>
        <configuration key="source-flavor">*/source</configuration>
        <configuration key="target-flavor">*/crop</configuration>
      </configurations>
    </operation>
```

But the target flavor option does not exist. This means that you always end up with multiple tracks with the same flavor making it harder to construct the rest of the workflow. Full example workflow: https://data.lkiesow.de/opencast/crop-test.xml

Another unexpected behavior is that this video which contains completely black letterboxing is not cropped:
https://data.lkiesow.de/opencast/sintel-trailer-letterbox.mp4

Commit:
target flavors added

isabellakutger commented:
@lkiesow I added the target flavors. Now, your workflow is working. 

The workflow operation does not crop letterboxing. It only checks the right and left side of  a track for black areas and crops them.

Commit:
version in pom.xml updated

staubesv commented:
@lkiesow Just wondering: Are there still open comments that @isabellakutger needs to resolve?

KatrinIhler commented:
@lkiesow Reminder! ;)

lkiesow commented on modules/crop-ffmpeg/src/test/java/org/opencastproject/crop/CropServiceTest.java:
```diff
@@ -0,0 +1,207 @@
+  /** Video file to test. Has black bars on each side */
+  protected static final String mediaResource = "/black.mp4";
+
+  /** Video file to test entire black slides. Shoudn't not be cropped */
```
```suggestion
  /** Video file to test entire black slides. Should not be cropped */
```

lkiesow commented on modules/crop-ffmpeg/src/test/java/org/opencastproject/crop/CropServiceTest.java:
```diff
@@ -0,0 +1,207 @@
+import org.opencastproject.workspace.api.Workspace;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.tika.io.IOUtils;
```
This does not define any Tika dependency and tika is not included by default anymore.
Apache commons-io does include `IOUtils` though.
Without this change, the patch will currently not built if merged into `develop`.

```suggestion
import org.apache.commons.io.IOUtils;
```

lkiesow commented on modules/crop-remote/src/main/java/org/opencastproject/crop/remote/CropServiceRemoteImpl.java:
```diff
@@ -0,0 +1,81 @@
+  /** The logger */
+  private static final Logger logger = LoggerFactory.getLogger(CropServiceRemoteImpl.class);
+
+  protected CropServiceRemoteImpl() {
```
Just starting this service will give you an error:

```
2019-02-13 15:34:17,268 | ERROR | (PaxLoggingServiceImpl$1ManagedPaxLoggingService:398) - [org.opencastproject.crop.remote.CropServiceRemoteImpl(79)] Error during instantiation of the implementation object
java.lang.IllegalAccessException: Class org.apache.felix.scr.impl.manager.SingleComponentManager can not access a member of class org.opencastproject.crop.remote.CropServiceRemoteImpl with modifiers "protected"
        at sun.reflect.Reflection.ensureMemberAccess(Reflection.java:102) ~[?:?]
        at java.lang.Class.newInstance(Class.java:436) ~[?:?]
        at org.apache.felix.scr.impl.manager.SingleComponentManager.createImplementationObject(SingleComponentManager.java:237) ~[?:?]
…
```

lkiesow commented on modules/crop-ffmpeg/src/main/resources/OSGI-INF/crop.xml:
```diff
@@ -0,0 +1,37 @@
+    <implementation class="org.opencastproject.crop.endpoint.CropRestEndpoint" />
+    <property name="service.description" value="Video Crop REST Endpoint" />
+    <property name="opencast.service.type" value="org.opencastproject.crop" />
+    <property name="opencast.service.path" value="/analysis/crop" />
```
The remote operation fails which is no wonder since this path does not match the path used in the remote operation:

```
2019-02-15 14:16:53,954 | WARN  | (RemoteBase:224) - Service at http://localhost:8081/analysis/crop/crop returned unexpected response code 405                                                                    
2019-02-15 14:16:53,955 | WARN  | (RemoteBase:241) - All services of type 'org.opencastproject.crop' are in unknown state, abort remote call /crop                                                                
2019-02-15 14:16:53,955 | ERROR | (WorkflowOperationWorker:180) - Workflow operation 'operation:'crop-video', position:5, state:'FAILED'' failed                                                                  
org.opencastproject.crop.api.CropException: Unable to crop trackhttp://localhost:8080/files/mediapackage/c1bcc318-ef4b-4ea7-ad4f-be6e3ae5bdd2/af0e9771-8bef-4afc-9128-2febde1ca5c8/test.mp4 using remote crop servi
ce
        at org.opencastproject.crop.remote.CropServiceRemoteImpl.crop(CropServiceRemoteImpl.java:79) ~[?:?]
        at org.opencastproject.workflow.handler.crop.CropWorkflowOperationHandler.start(CropWorkflowOperationHandler.java:117) ~[?:?]
        at org.opencastproject.workflow.impl.WorkflowOperationWorker.start(WorkflowOperationWorker.java:233) ~[177:opencast-workflow-service-impl:7.0.0.SNAPSHOT]
        at org.opencastproject.workflow.impl.WorkflowOperationWorker.execute(WorkflowOperationWorker.java:155) [177:opencast-workflow-service-impl:7.0.0.SNAPSHOT]
        at org.opencastproject.workflow.impl.WorkflowServiceImpl.runWorkflowOperation(WorkflowServiceImpl.java:857) [177:opencast-workflow-service-impl:7.0.0.SNAPSHOT]
        at org.opencastproject.workflow.impl.WorkflowServiceImpl.process(WorkflowServiceImpl.java:1890) [177:opencast-workflow-service-impl:7.0.0.SNAPSHOT]
        at org.opencastproject.workflow.impl.WorkflowServiceImpl$JobRunner.call(WorkflowServiceImpl.java:2318) [177:opencast-workflow-service-impl:7.0.0.SNAPSHOT]
        at org.opencastproject.workflow.impl.WorkflowServiceImpl$JobRunner.call(WorkflowServiceImpl.java:2284) [177:opencast-workflow-service-impl:7.0.0.SNAPSHOT]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
        at java.lang.Thread.run(Thread.java:748) [?:?]
```

lkiesow committed:
MH-12821, Do not import Tika

The crop service imports classes from Apache but does not specify this
dependency causing the vuild to fail since other components of Opencast
do not use it anymore and so it is not accidentally there any longer.

lkiesow committed:
MH-12821, Crop Operation Documentation

This patch adds documentation for the supported workflow operation
configuration options.

lkiesow committed:
MH-12821, Fix Remote Crop Service Activation

This patch makes the constructor of the remote crop service public since
the activation of the OSGI bundle will fail otherwise. For example,
without this, you cannot properly start an admin node.

lkiesow committed:
MH-12821, Fix Crop Remote Service Path

This patch fixes REST endpoint location for the crop service since the
path used in the REST endpoint and in the remote service do not match
causing the service to always fail.

lkiesow reviewed (changes requested):
I'did a new review but still found this pull request non-functional. Specifically on distributed systems.
I've submitted a few patches  to already fix a number of problems but there is still at least one remaining which causes the operation to fail on a distributed system with non-shared storage:

```
2019-02-15 16:40:29,542 | ERROR | (WorkflowOperationWorker:182) - Workflow operation 'operation:'crop-video', position:5, state:'FAILED'' failed                                                                  
org.opencastproject.workflow.api.WorkflowOperationException: Could not move http://localhost:8081/files/collection/cropping/a502c787_3f03_4532_a76e_a098adc84faf_sintel_trailer_1080p.mp4 to media package 334e60a1
-dd62-468e-accc-d19a154c863b                                                                                                                                                                                      
        at org.opencastproject.workflow.handler.crop.CropWorkflowOperationHandler.start(CropWorkflowOperationHandler.java:149) ~[?:?]                                                                             
        at org.opencastproject.workflow.impl.WorkflowOperationWorker.start(WorkflowOperationWorker.java:233) ~[177:opencast-workflow-service-impl:7.0.0.SNAPSHOT]                                                 
        at org.opencastproject.workflow.impl.WorkflowOperationWorker.execute(WorkflowOperationWorker.java:155) [177:opencast-workflow-service-impl:7.0.0.SNAPSHOT]                                                
        at org.opencastproject.workflow.impl.WorkflowServiceImpl.runWorkflowOperation(WorkflowServiceImpl.java:857) [177:opencast-workflow-service-impl:7.0.0.SNAPSHOT]                                           
        at org.opencastproject.workflow.impl.WorkflowServiceImpl.process(WorkflowServiceImpl.java:1890) [177:opencast-workflow-service-impl:7.0.0.SNAPSHOT]                                                       
        at org.opencastproject.workflow.impl.WorkflowServiceImpl$JobRunner.call(WorkflowServiceImpl.java:2318) [177:opencast-workflow-service-impl:7.0.0.SNAPSHOT]                                                
        at org.opencastproject.workflow.impl.WorkflowServiceImpl$JobRunner.call(WorkflowServiceImpl.java:2284) [177:opencast-workflow-service-impl:7.0.0.SNAPSHOT]                                                
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]                                                                                                                                         
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]                                                                                                                  
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]                                                                                                                  
        at java.lang.Thread.run(Thread.java:748) [?:?]                                                                                                                                                            
2019-02-15 16:40:29,584 | ERROR | (WorkflowOperationWorker:182) - Workflow operation 'operation:'crop-video', position:5, state:'FAILED'' failed                                                                  
org.opencastproject.workflow.api.WorkflowOperationException: Could not move http://localhost:8081/files/collection/cropping/ae59aa8c_5f6e_49f6_804f_15721e09b3fc_sintel_trailer_1080p.mp4 to media package 8e4591b9
-c4cb-4533-ac1d-4b1788496394                                                                                                                                                                                      
        at org.opencastproject.workflow.handler.crop.CropWorkflowOperationHandler.start(CropWorkflowOperationHandler.java:149) ~[?:?]                                                                             
        at org.opencastproject.workflow.impl.WorkflowOperationWorker.start(WorkflowOperationWorker.java:233) ~[177:opencast-workflow-service-impl:7.0.0.SNAPSHOT]                                                 
        at org.opencastproject.workflow.impl.WorkflowOperationWorker.execute(WorkflowOperationWorker.java:155) [177:opencast-workflow-service-impl:7.0.0.SNAPSHOT]                                                
        at org.opencastproject.workflow.impl.WorkflowServiceImpl.runWorkflowOperation(WorkflowServiceImpl.java:857) [177:opencast-workflow-service-impl:7.0.0.SNAPSHOT]                                           
        at org.opencastproject.workflow.impl.WorkflowServiceImpl.process(WorkflowServiceImpl.java:1890) [177:opencast-workflow-service-impl:7.0.0.SNAPSHOT]                                                       
        at org.opencastproject.workflow.impl.WorkflowServiceImpl$JobRunner.call(WorkflowServiceImpl.java:2318) [177:opencast-workflow-service-impl:7.0.0.SNAPSHOT]                                                
        at org.opencastproject.workflow.impl.WorkflowServiceImpl$JobRunner.call(WorkflowServiceImpl.java:2284) [177:opencast-workflow-service-impl:7.0.0.SNAPSHOT]                                                
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]                                                                                                                                         
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]                                                                                                                  
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]                                                                                                                  
        at java.lang.Thread.run(Thread.java:748) [?:?] 
```

My guess would be that the location of the files of the workers is put into the tracks as location but then they are not properly accessible from the admin node. But I did not investigate this in more detail.

The test setup is basically:

- admin running on localhost:8080
- presentation running on localhost:8081
- worker running on localhost:8082
- all connected to a local MariaDB
- `fast` workflow with crop included

Everything else is left at the default settings.

KatrinIhler commented:
@isabellakutger Is there still work being done on this?

isabellakutger commented:
@KatrinIhler it's paused till July because I'm currently working on a other project

KatrinIhler commented:
I'm gonna close this for now. Please reopen this once further work has been done and you feel confident that the problems are fixed and it can be merged.
